exiftime.1
.TH EXIFTIME 1
.\"
.\" Copyright (c) 2004-2007, Eric M. Johnston <emj@postal.net>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"      This product includes software developed by Eric M. Johnston.
.\" 4. Neither the name of the author nor the names of any co-contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. 
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\" $Id: exiftime.1,v 1.2 2007/12/16 02:19:25 ejohnst Exp $
.\"
.SH NAME
.B exiftime
\- display or adjust date & time Exif tags; list files ordered by their
Exif date & time tags
.SH SYNOPSIS
.B exiftime
.RB [ \-filqw ]
.RB [ \-s
.IR delim ]
.RB [ \-t [ acdg ]]
.RB [ \-v [ + | \- ] \fIval [ ymwdHMS ]]
.I file ...
.SH DESCRIPTION
When invoked without arguments, the
.B exiftime
utility displays the Exif date and time tags contained in each input
.I file
to the standard output.  Otherwise, depending on the options specified,
.B exiftime
will operate on only the chosen tags, will copy one tag's time to others,
will adjust the date and time, will write an adjusted time to each
.IR file ,
or will list each
.I file
in ascending order by date and time.

Most digital cameras include one or more date and time tags in the Exif
data added to the image files they produce.  These tags are:
.IP "Image Created" 4
The date and time the image was created or changed.  This is the most common tag.
.IP "Image Generated" 4
The date and time the original image data was generated (i.e., when picture was taken).
.IP "Image Digitized" 4
The date and time the image was stored as digital data.
.PP
The format for these tags is "YYYY:MM:DD HH:MM:SS" with the time shown in
24-hour format.  The
.B exiftime
utility cannot add a tag if it does not already exist in
.IR file .

By default,
.B exiftime
will simply print out any of the three date and time tags in each input
.IR file .
The
.B -l
flag
will produce a listing of each input
.I file
in date/time order, suitable for use in
further image processing (e.g., production of a Web catalog) when filename
ordering is not useful.

The
.B -c
flag may be used to copy one tag's time to other tags you specify.  By default,
it just prints the copied timestamp, but is written to the file when used
with the
.B -w
flag.

The
.B -v
flag may be used to vary, or adjust, dates and times.  When used with the
.B -w
flag, which writes the adjusted date and time to each input
.IR file ,
one may, for example, process a batch of files to adjust for a camera's
incorrectly set clock.
.SH OPTIONS
.IP -c
Select the source date and time tag to copy when followed by one of
.B c
(Image Created),
.B d
(Image Digitized), or 
.B g
(Image Generated). This source tag's value is copied to one or more other tags
specified using the
.B -t
flag. The
.B -v
flag cannot be used when copying a tag.
.IP -f
Write adjusted date and time tags without prompting for confirmation.
.IP -i
Output a prompt to standard error before overwriting a date and time tag with
the adjusted value.  If the response from the standard input begins with 'y'
or 'Y', the tag is overwritten.  This option is default behavior.
.IP -l
List each input
.I file
in ascending order by timestamp.  By default, it uses the Image Created tag.
In the absence of an Image Created tag, first Image Generated then Image
Digitized are used.  Alternatively, the
.B -t
flag may be used to specify the timestamp preference for ordering.  If no
date and time tags are present, the file's modified time is used.  This flag
overrides all others but the
.B -t
flag.
.IP -q
Do not output details of a date and time adjustment to standard out when
using the
.B -w
flag.
.IP -s
Separate field name and value with the string
.IR delim  .
The default is ': '.
.IP -t
Select the date and time tags for display or adjustment when followed by one
or more of
.B a
(all tags),
.B c
(Image Created),
.B d
(Image Digitized), or 
.B g
(Image Generated).
.IP -v
Adjust the date and time tags' second, minute, hour, month day, week day,
month or year according to
.IR val .
If
.I val
is preceded with a plus or minus sign, the date is adjusted forwards or
backwards according to the remaining string; otherwise the relevant part of
the date is set.  The date can be adjusted as many times as required using
these flags.  Flags are processed in the order given.

When providing an absolute value (rather than a relative adjustment),
seconds are in the range 0-59, minutes are in the range 0-59, hours are
in the range 0-23, month days are in the range 1-31, week days are in the
range 0-6 (Sun-Sat), months are in the range 1-12 (Jan-Dec) and years are in
the range 80-38 or 1980-2038.

If
.I val
is numeric, one of either
.BR y ,
.BR m ,
.BR w ,
.BR d ,
.BR H ,
.BR M ", or "
.B S
must be used to specify which part of the date is to be adjusted.

The week day or month may be specified using a name rather than a number.
If a name is used with the plus (or minus) sign, the date will be put forwards
(or backwards) to the next (previous) date that matches the given week day
or month.  This will not adjust the date, if the given week day or month is
the same as the current one.

When the date is adjusted to a specific value that doesn't actually exist
(for example March 26, 1:30 BST 2000 in the Europe/London timezone),
the date will be silently adjusted forwards in units of one hour until it
reaches a valid time.  When the date is adjusted to a specific value that
occurs twice (for example October 29, 1:30 2000), the resulting timezone will
be set so that the date matches the earlier of the two times.  In all cases,
daylight savings time considerations are ignored.

Refer to the examples below for further details.
.IP -w
Write the adjusted or copied date and time tags.  By default, any of the
three date and time tags present in the file are adjusted; otherwise, only
those specified with the
.B -t
flag are adjusted or copied.
.SH EXAMPLES
The command
.IP
.nf
exiftime example1.jpg
.fi
.PP
will display:
.IP
.nf
Image Created: 2003:09:12 17:05:37
Image Generated: 2003:09:12 17:05:37
Image Digitized: 2003:09:12 17:05:37
.fi
.PP
The command
.IP
.nf
exiftime -tcd example1.jpg
.fi
.PP
will display:
.IP
.nf
Image Created: 2003:09:12 17:05:37
Image Digitized: 2003:09:12 17:05:37
.fi
.PP
The command
.IP
.nf
exiftime -cg -tc -w example1.jpg
.fi
.PP
will copy the Image Generated tag's date and time to the Image Created tag and
write the result, prompting the user to first confirm the operation:
.IP
.nf
adjust time created in example1.jpg from
  2011:09:29 10:38:11 to 2011:08:27 21:23:50? (y/n [n]) y
Image Created: 2011:09:29 10:38:11 -> 2011:08:27 21:23:50
.fi
.PP
The command
.IP
.nf
exiftime -v+3H example1.jpg
.fi
.PP
will adjust each time forward by three hours and display:
.IP
.nf
Image Created: 2003:09:12 20:05:37
Image Generated: 2003:09:12 20:05:37
Image Digitized: 2003:09:12 20:05:37
.fi
.PP
The command
.IP
.nf
exiftime -v+5d -v-7M -fw -tg *.jpg
.fi
.PP
will adjust the date ahead five days and the time back seven minutes and
write the adjusted date and time to the Image Generated tag without a prompt
for confirmation for all files that match "*.jpg".  It displays:
.IP
.nf
example1.jpg:
Image Generated: 2003:09:12 17:05:37 -> 2003:09:17 16:58:37

example2.jpg:
Image Generated: 2004:01:22 17:07:02 -> 2004:01:27 17:00:02
.fi
.PP
The command
.IP
.nf
exiftime -l -tdg *.jpg
.fi
.PP
will list all files that match "*.jpg", one per line, in ascending timestamp
order.  It'll attempt to use the following timestamp values, in order: Image
Digitized, Image Generated, Image Created, and, finally, the OS's epoch.
.SH DIAGNOSTICS
The
.B exiftime
utility exits 0 on success and 1 if an error occurs.
.SH "SEE ALSO"
exiftags(1), exifcom(1)
.SH STANDARDS
The
.B exiftime
utility was developed using the 2003 draft Exif standard, version 2.21
(http://tsc.jeita.or.jp/).
.SH BUGS
Does not support the Exif tags SubsecTime, SubsecTimeOriginal, or
SubsecTimeDigitized.  Does not support manufacturer-specific date and time
tags.
.SH AUTHOR
The
.B exiftime
utility and this man page were written by Eric M. Johnston <emj@postal.net>.
The time adjustment functionality and documentation were derived from portions
of FreeBSD's date(1) command by Brian Somers <brian@Awfulhak.org>.


CHANGES
20130704 use file modified time as backup for listing in exiftime
20130704 added timestamp copy to exiftime
20071215 version: exiftags 1.01
20071215 fixed to compile on HP-UX
20071215 added Canon EOS-5D custom function support (thanks: Albert Max Lai)
20071215 custom timestamp prefs in exiftime listing (thanks: Stuart Scharf)
20071215 suppress warnings on invalid user comments
20071214 allow output of invalid tags in debug
20071214 fixed looping from recursive IFD ref, resolves CVE-2007-6356
         (thanks: Christian Schmid; Meder Kydyraliev, Google Security Team)
20071214 fixed handling of field offset overflow, resolves CVE-2007-6354 &
         CVE-2007-6355 (thanks: Meder Kydyraliev, Google Security Team)

20050126 version: exiftags 1.00
20050107 added Canon tags, minor tweaks (thanks: Henrik Levkowetz)
20050104 fixed handling of short ASCII strings
20041228 added Panasonic tags (thanks: Lee Kindness)
20041227 enhanced Nikon maker note support for newer cameras
20041223 internal code cleanup to allow separate maker TIFF info
20041220 added new Canon shooting modes (thanks: Henrik Levkowetz)
20041217 fixed exiftime crash on missing tags (thanks: Nathan Treadway)
20041106 fixed looping from recursive Canon IFD ref (thanks: Mathieu Vanasse)
20041011 sanitize ASCII tags to handle corrupt maker notes
20041010 added Panasonic values, tags (thanks: Laurent Monin)
20041009 image number support for Canon EOS-20D (thanks: Per Kristian Hove)
20041009 added Canon EOS-20D custom function support
20040915 added Sigma maker note support
20040821 added Canon tag values, calculate EV's (thanks: Bradley Bailey)
20040820 fixed flash tag processing to catch all possible values
20040820 fixed spelling for Minolta values (thanks: Martin Pool)

20040506 version: exiftags 0.99.1
20040506 fixed exiftime install oversight
20040504 added Canon camera actuations (thanks: Stan Jirman)

20040430 version: exiftags 0.99
20040420 added Panasonic maker note support (thanks: Tom Hughes)
20040410 improved error messages in exifcom
20040408 added exiftime utility to display and write date/time tags
20040403 added Leica maker note support
20040403 slight hack to use Canon's subject distance tag (units unclear)
20040403 fixed Canon maker function parameter mismatch
20040403 updated for Exif 2.21 draft spec (minor changes)
20040403 fixed problem with images without IFD1
20040403 additional bounds checks on some tags (thanks: Michael McCandless)
20040403 fixed negative fraction display
20040227 added Asahi/Pentax maker note support
20031101 prevent looping when a tag refers to its own IFD

20030815 version: exiftags 0.98
20030807 fixed some Fuji, Minolta, Nikon byte order bugs/dependencies
20030806 internal code cleanup to nicely support multiple tag sets
20030806 internal code cleanup to allow different IFD byte orders
20030806 added Sanyo maker note support
20030805 added support for Nikon D1X
20030805 fixed crash on Minolta DiMAGE EX
20030802 fixed offset check bug for ascii strings
20030801 added support for GPS info tags
20030622 fixed crashes on badly formed comments, type mismatches
20030622 added Canon 10D custom function support (thanks: Jason Montojo)
20030211 tweaked Canon 1D/1Ds custom function support

20030204 version: exiftags 0.97
20030204 fixed byte order bug displaying Exif version
20030128 assign bad or senseless tags separate dump level
20030124 internal code cleanups
20030123 assign overridden tags separate dump level
20030121 support for Canon EOS-1D & 1DS (thanks: Stan Jirman)
20030120 added -s option to customize delimiter
20030120 support for multiple shorts in a tag (e.g., subject area)
20030112 added Minolta maker note support (thanks: Javier Crespo)
20030104 removed spurious Exif header warning (thanks: Kevin Layer)
20030104 Canon ISO values (thanks: Alan Murphy)

20021103 version: exiftags 0.96
20021103 internal mechanism for one tag to override another
20021102 reduce fractions (e.g., exposure time)
20021102 updated for Exif 2.2 spec: new tags, some existing tags expanded
20021102 made flash tag support safer (thanks: Christian Zuckschwerdt)
20021102 save Exif version with file info
20021101 added Casio maker note support
20021101 fixed segfault on bad Exif data

20021006 version: exiftags 0.95
20021007 added exifcom utility to display and write UserComment
20021006 fixed read of max len UserComment
20021006 internal code cleanup to allow additional utilities
20021006 made ISO setting image-specific
20021005 internal change to value description lookups
20021005 added support for Canon D30 (and D60?) custom functions
20021005 added new Canon maker note tags
20021005 added -l option for removable lens cameras
20021004 made shutter speed verbose when exposure time available
20021004 smarter rational type handling

20020928 version: exiftags 0.94
20020928 support ASCII UserComment (thanks: Alan Judge)

20020831 version: exiftags 0.93
20020831 print warnings instead of dying on some errors (thanks: Ian Morgan)

20020801 version: exiftags 0.92
20020731 prevent potential NaN values
20020731 portability changes for Windows

20020712 version: exiftags 0.91
20020711 portability changes for Solaris
20020711 modifications too numerous to mention since exifdump
20020630 changed name to exiftags

20020216 version: exifdump 0.90
20020120 version: exifdump 0.80

$Id: CHANGES,v 1.23 2007/12/16 03:45:59 ejohnst Exp $


exiftime.c
/*
 * Copyright (c) 2004-2007, Eric M. Johnston <emj@postal.net>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *      This product includes software developed by Eric M. Johnston.
 * 4. Neither the name of the author nor the names of any co-contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * $Id: exiftime.c,v 1.13 2007/12/16 02:18:51 ejohnst Exp $
 */

/*
 * exiftime: display or adjust date & time Exif tags; list files
 * ordered by time.
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>

/* For getopt(). */

#ifndef WIN32
#include <unistd.h>
#else
extern char *optarg;
extern int optind, opterr, optopt;
int getopt(int, char * const [], const char *);
#endif

#include "jpeg.h"
#include "exif.h"
#include "timevary.h"


struct linfo {
	const char *fn;
	time_t ts;
};

static const char *version = "1.02";
static int iflag, lflag, qflag, wflag, ttags, ctags;
static const char *delim = ": ";
static const char *fname;
static struct vary *v;
static struct linfo *lorder;

#define EXIFTIMEFMT	"%Y:%m:%d %H:%M:%S"
#define EXIFTIMELEN	20

#define ET_CREATE	0x01
#define ET_GEN		0x02
#define ET_DIGI		0x04


/*
 * Some helpful info...
 */
static
void usage()
{
	fprintf(stderr, "Usage: %s [options] file ...\n", progname);
	fprintf(stderr, "Displays or adjusts Exif timestamps in the specified "
	    "files.\n");
	fprintf(stderr, "Version: %s\n\n", version);
	fprintf(stderr, "Available options:\n");
	fprintf(stderr, "  -f\tAnswer 'yes' to any confirmation prompts.\n");
	fprintf(stderr, "  -i\tConfirm overwrites (default).\n");
	fprintf(stderr, "  -l\tList files ordered by timestamp, using image "
	    "created by default;\n\toverrides -v, -w.  Use -t to change "
	    "timestamp preference.\n");
	fprintf(stderr, "  -q\tBe quiet when writing timestamps.\n");
	fprintf(stderr, "  -s\tSet delimiter to provided string "
	    "(default: \": \").\n");
	fprintf(stderr, "  -t[acdg]\n\tDisplay/adjust specified timestamp(s), "
	    "if they exist:\n");
	fprintf(stderr, "\t  a: All available timestamps (default);\n\t  "
	    "c: Image created;\n\t  g: Image generated; or\n\t  d: Image "
	    "digitized.\n");
	fprintf(stderr, "  -v[+|-]val[ymwdHMS]\n\tAdjust the timestamp(s) by "
	    "the given amount.\n");
	fprintf(stderr, "  -c[c|d|g]\n\tCopy the timestamp to those "
	    "specified by -t.\n");
	fprintf(stderr, "  -w\tWrite adjusted or copied timestamp(s).\n");

	vary_destroy(v);
	exit(1);
}


/*
 * Compare two linfo members for our sort.
 */
int
lcomp(const void *a, const void *b)
{

	if (((struct linfo *)a)->ts < ((struct linfo *)b)->ts)
		return (-1);
	if (((struct linfo *)a)->ts == ((struct linfo *)b)->ts)
		return (0);
	return (1);
}


/*
 * Stuff a standard Exif timestamp into a struct *tm.
 * I've got to say that it's pretty annoying that Win32 doesn't have
 * strptime()...
 */
static int
etstrptm(const char *buf, struct tm *tp)
{
	int n;

	memset(tp, 0, sizeof(struct tm));
	n = sscanf(buf, "%d:%d:%d %d:%d:%d", &tp->tm_year, &tp->tm_mon,
	    &tp->tm_mday, &tp->tm_hour, &tp->tm_min, &tp->tm_sec);
	tp->tm_year -= 1900;
	tp->tm_mon -= 1;

	return (n != 6);
}


/*
 * Grab the timestamps for listing in the specified order of preference.
 * Doesn't modify the file.
 */
static int
listts(FILE *fp, struct exiftags *t, struct linfo *li, u_int16_t *tpref)
{
	struct exifprop *p;
	struct tm tv;
	struct stat finfo;

	/* If no timestamp is found, print error and list first. */

	p = findprop(t->props, tags, tpref[0]);

	if (!p || !p->str || etstrptm(p->str, &tv)) {
		p = findprop(t->props, tags, tpref[1]);

		if (!p || !p->str || etstrptm(p->str, &tv)) {
			p = findprop(t->props, tags, tpref[2]);

			if (!p || !p->str || etstrptm(p->str, &tv)) {
				exifwarn("no timestamp available; using mtime");
				fstat(fileno(fp), &finfo);
				li->ts = finfo.st_mtime;
				return (1);
			}
		}
	}

	li->ts = mktime(&tv);
	return (0);
}


/*
 * Grab the specified timestamp and vary it, if necessary.
 * The provided buffer must be at least EXIFTIMELEN bytes.
 */
static int
ettime(char *b, struct exifprop *p)
{
	struct tm tv;
	const struct vary *badv;

	/* Slurp the timestamp into tv. */

	if (!p || !p->str || etstrptm(p->str, &tv))
		return (1);

	/* Apply any adjustments.  (Bad adjustment = fatal.) */

	badv = vary_apply(v, &tv);
	if (badv) {
		exifwarn2("cannot apply adjustment", badv->arg);
		usage();
	}

	if (strftime(b, EXIFTIMELEN, EXIFTIMEFMT, &tv) != EXIFTIMELEN - 1)
		return (1);

	return (0);
}


/*
 * Overwrite a timestamp with the adjusted value.
 */
static int
writets(FILE *fp, long pos, struct exiftags *t, struct exifprop *p,
    const unsigned char *buf, const char *ttype, const char *nts)
{
	int ch, checkch;
	long psave;

	/* Some sanity checking. */

	if (strlen(nts) != EXIFTIMELEN - 1) {
		fprintf(stderr, "%s: invalid timestamp -- %s\n", fname, nts);
		return (1);
	}

	if (!strcmp(nts, p->str)) {
		fprintf(stderr, "%s: new %s timestamp identical to old\n",
		    fname, ttype);
		return (1);
	}

	/* Prompt user, if desired. */

	if (iflag) {
		fprintf(stderr, "adjust time %s in %s from\n  %s to %s? "
		    "(y/n [n]) ", ttype, fname, p->str, nts);
		checkch = ch = getchar();
		while (ch != '\n' && ch != EOF)
			ch = getchar();
		if (checkch != 'y' && checkch != 'Y') {
			fprintf(stderr, "not adjusted\n");
			return (1);
		}
	}

	/* Remember where we are and move to the comment in our file. */

	psave = ftell(fp);
	if (fseek(fp, pos + (t->md.btiff - buf) + p->value, SEEK_SET))
		exifdie((const char *)strerror(errno));

	/* Write the new timestamp. */

	if (fwrite(nts, EXIFTIMELEN, 1, fp) != 1)
		exifdie((const char *)strerror(errno));

	/* Restore the file pointer. */

	if (fseek(fp, psave, SEEK_SET))
		exifdie((const char *)strerror(errno));

	if (!qflag)
		printf("%s%s%s -> %s\n", p->descr, delim, p->str, nts);
	return (0);
}


/*
 * Process a timestamp.
 * Returns 0 on success, 1 if it had trouble writing, 2 if the tag wasn't
 * found (and it was explictly requested), or -1 if the tag wasn't found.
 */
static int
procts(FILE *fp, long pos, struct exiftags *t, const unsigned char *buf,
    u_int16_t tag, const char *ttype)
{
	int rc;
	char nts[EXIFTIMELEN];
	struct exifprop *p;

	p = findprop(t->props, tags, tag);
	if (ettime(nts, p)) {

		/*
		 * If ttags != 0, then the user explicitly requested the
		 * timestamp, so print an error if it doesn't exist.
		 */
		if (ttags) {
			fprintf(stderr, "%s: image %s time not available\n",
			    fname, ttype);
			return (2);
		}

		return (-1);
	}

	if (wflag)
		rc = writets(fp, pos, t, p, buf, ttype, nts);
	else {
		printf("%s%s%s\n", p->descr, delim, nts);
		rc = 0;
	}

	return (rc);
}


/*
 * Copy a timestamp to the specified timestamp.
 * Returns 0 on success, 1 if it had trouble writing or tags weren't found.
 */
static int
copyts(FILE *fp, long pos, struct exiftags *t, const unsigned char *buf,
    u_int16_t tag, const char *ttype, const char *nts)
{
	int rc;
	struct exifprop *p;

	p = findprop(t->props, tags, tag);
	if (!p) {
		fprintf(stderr, "%s: image %s time not available\n",
		    fname, ttype);
		return (1);
	}

	if (wflag)
		return (writets(fp, pos, t, p, buf, ttype, nts));

	printf("%s%s%s\n", p->descr, delim, nts);
	return (0);
}


/*
 * Copy timestamp into specified timestamps.
 * Returns 0 on success, 1 if it had trouble writing or tags weren't found.
 */
static int
copyallts(FILE *fp, long pos, struct exiftags *t, const unsigned char *buf,
    u_int16_t tag, const char *ttype)
{
	int r, rc;
	char nts[EXIFTIMELEN];
	struct exifprop *p;

	p = findprop(t->props, tags, tag);
	if (ettime(nts, p)) {
		fprintf(stderr, "%s: image %s time not available\n",
		    fname, ttype);
		return (1);
	}

	if (ttags & ET_CREATE) {
		r = copyts(fp, pos, t, buf, EXIF_T_DATETIME, "created", nts);
		if (r > 0) rc = 1;
	}

	if (ttags & ET_GEN) {
		r = copyts(fp, pos, t, buf, EXIF_T_DATETIMEORIG,
		    "generated", nts);
		if (r > 0) rc = 1;
	}

	if (ttags & ET_DIGI) {
		r = copyts(fp, pos, t, buf, EXIF_T_DATETIMEDIGI,
		    "digitized", nts);
		if (r > 0) rc = 1;
	}

	return (rc);
}


/*
 * Process the file's timestamps according to the options chosen.
 */
static int
procall(FILE *fp, long pos, struct exiftags *t, const unsigned char *buf)
{
	int r, rc, found;

	found = rc = 0;

	if (ctags) {
		switch (ctags) {
		case ET_CREATE:
			rc = copyallts(fp, pos, t, buf, EXIF_T_DATETIME,
			    "created");
			break;
		case ET_GEN:
			rc = copyallts(fp, pos, t, buf, EXIF_T_DATETIMEORIG,
			    "generated");
			break;
		case ET_DIGI:
			rc = copyallts(fp, pos, t, buf, EXIF_T_DATETIMEDIGI,
			    "digitized");
			break;
		default:
			fprintf(stderr, "%s: source timestamp not available\n",
			    fname);
			rc = 1;
		}
		return (rc);
	}

	/*
	 * If ttags = 0, process them all or an error if there are none.
	 */

	if (ttags & ET_CREATE || !ttags) {
		r = procts(fp, pos, t, buf, EXIF_T_DATETIME, "created");
		if (r == 0 || r == 1) found++;
		if (r > 0) rc = 1;
	}

	if (ttags & ET_GEN || !ttags) {
		r = procts(fp, pos, t, buf, EXIF_T_DATETIMEORIG, "generated");
		if (r == 0 || r == 1) found++;
		if (r > 0) rc = 1;
	}

	if (ttags & ET_DIGI || !ttags) {
		r = procts(fp, pos, t, buf, EXIF_T_DATETIMEDIGI, "digitized");
		if (r == 0 || r == 1) found++;
		if (r > 0) rc = 1;
	}

	/* No timestamp tags. */

	if (!ttags && !found) {
		fprintf(stderr, "%s: no timestamps available\n", fname);
		return (1);
	}

	return (rc);
}


/*
 * Scan the JPEG file for Exif data and parse it.
 */
static int
doit(FILE *fp, int n, u_int16_t *tpref)
{
	int mark, gotapp1, first, rc;
	unsigned int len, rlen;
	unsigned char *exifbuf;
	struct exiftags *t;
	long app1;

	gotapp1 = FALSE;
	first = 0;
	exifbuf = NULL;
	rc = 0;

	while (jpegscan(fp, &mark, &len, !(first++))) {

		if (mark != JPEG_M_APP1) {
			if (fseek(fp, len, SEEK_CUR))
				exifdie((const char *)strerror(errno));
			continue;
		}

		exifbuf = (unsigned char *)malloc(len);
		if (!exifbuf)
			exifdie((const char *)strerror(errno));

		app1 = ftell(fp);
		rlen = fread(exifbuf, 1, len, fp);
		if (rlen != len) {
			fprintf(stderr, "%s: error reading JPEG (length "
			    "mismatch)\n", fname);
			free(exifbuf);
			return (1);
		}

		t = exifscan(exifbuf, len, FALSE);

		if (t && t->props) {
			gotapp1 = TRUE;
			if (lflag)
				rc = listts(fp, t, &lorder[n], tpref);
			else
				rc = procall(fp, app1, t, exifbuf);
		}
		exiffree(t);
		free(exifbuf);
	}

	if (!gotapp1) {
		fprintf(stderr, "%s: couldn't find Exif data\n", fname);
		return (1);
	}

	return (rc);
}


/*
 * Fill in array of timestamps in order of preference.
 */
void
settpref(u_int16_t *tpref, u_int16_t tag)
{
	int i;

	/* Recursively fill in the remaining timestamps. */

	if (tag == EXIF_T_UNKNOWN) {
		settpref(tpref, EXIF_T_DATETIME);
		settpref(tpref, EXIF_T_DATETIMEORIG);
		settpref(tpref, EXIF_T_DATETIMEDIGI);
		return;
	}

	for (i = 0; i < 3; i++) {
		if (tpref[i] == tag)
			break;
		if (tpref[i] == EXIF_T_UNKNOWN) {
			tpref[i] = tag;
			break;
		}
	}
}


int
main(int argc, char **argv)
{
	register int ch;
	int eval, fnum, wantall;
	char *rmode, *wmode;
	FILE *fp;
	u_int16_t tpref[3];

	progname = argv[0];
	debug = FALSE;
	ttags = ctags = wantall = eval = 0;
	lflag = qflag = wflag = FALSE;
	iflag = TRUE;
	v = NULL;
	tpref[0] = tpref[1] = tpref[2] = EXIF_T_UNKNOWN;
#ifdef WIN32
	rmode = "rb";
	wmode = "r+b";
#else
	rmode = "r";
	wmode = "r+";
#endif

	while ((ch = getopt(argc, argv, "filqs:t:c:v:w")) != -1)
		switch (ch) {
		case 'f':
			iflag = FALSE;
			break;
		case 'i':
			iflag = TRUE;
			break;
		case 'l':
			lflag = TRUE;
			break;
		case 'q':
			qflag = TRUE;
			break;
		case 's':
			delim = optarg;
			break;
		case 'c':
			if (strlen(optarg) != 1) {
				exifwarn("can only copy from one timestamp");
				usage();
			}
			switch (*optarg) {
			case 'c':
				ctags = ET_CREATE;
				break;
			case 'd':
				ctags = ET_DIGI;
				break;
			case 'g':
				ctags = ET_GEN;
				break;
			default:
				exifwarn("invalid timestamp for copy");
				usage();
			}
			break;
		case 't':
			while (*optarg) {
				switch (*optarg) {
				case 'c':
					settpref(tpref, EXIF_T_DATETIME);
					ttags |= ET_CREATE;
					break;
				case 'd':
					settpref(tpref, EXIF_T_DATETIMEDIGI);
					ttags |= ET_DIGI;
					break;
				case 'g':
					settpref(tpref, EXIF_T_DATETIMEORIG);
					ttags |= ET_GEN;
					break;
				case 'a':
					wantall = 1;
					break;
				default:
					exifwarn("invalid timestamp");
					usage();
				}
				optarg++;
			}
			if (wantall) ttags = 0;
			break;
		case 'v':
			v = vary_append(v, optarg);
			break;
		case 'w':
			wflag = TRUE;
			break;
		case '?':
		default:
			usage();
		}
	argc -= optind;
	argv += optind;

	if (!*argv)
		usage();

	if (v && ctags) {
		exifwarn("timestamp vary and copy are mutally exclusive");
		usage();
	}

	if (ctags & ttags) {
		exifwarn("can't copy timestamp to itself");
		usage();
	}

	if (ctags && !ttags) {
		exifwarn("need destination timestamp(s) to copy");
		usage();
	}

	/* Finish up timestamp preferences. */

	settpref(tpref, EXIF_T_UNKNOWN);

	/* Don't be quiet if we're not writing. */

	if (qflag && !wflag)
		qflag = FALSE;

	/* Prepare an array for sorting. */

	if (lflag) {
		wflag = 0;
		lorder = (struct linfo *)calloc(argc, sizeof(struct linfo));
		if (!lorder)
			exifdie((const char *)strerror(errno));
	}

	/* Run through the files... */

	for (fnum = 0; *argv; ++argv, fnum++) {

		fname = *argv;

		/* Only open for read+write if we need to. */

		if ((fp = fopen(*argv, wflag ? wmode : rmode)) == NULL) {
			exifwarn2(strerror(errno), *argv);
			eval = 1;
			continue;
		}

		/* Print filenames if more than one. */

		if (argc > 1 && !lflag && !qflag)
			printf("%s%s:\n", fnum == 0 ? "" : "\n", *argv);

		if (lflag)
			lorder[fnum].fn = fname;

		if (doit(fp, fnum, tpref))
			eval = 1;
		fclose(fp);
	}

	/*
	 * We'd like to use mergesort() here (instead of qsort()) because
	 * qsort() isn't stable w/members that compare equal and we exepect
	 * the list of files to frequently already be in order.  However,
	 * FreeBSD seems to be the only platform with mergesort(), and I'm
	 * not inclined to include the function...
	 */
	if (lflag) {
		qsort(lorder, argc, sizeof(struct linfo), lcomp);
		for (fnum = 0; fnum < argc; fnum++)
			printf("%s\n", lorder[fnum].fn);
		free(lorder);	/* XXX Over in usage()? */
	}

	vary_destroy(v);
	return (eval);
}
